More accurate conversion between latitude and authalic latitude

fgilic opened the pull request:
This pull request introduces changes listed below.

## Change 1 (5e2b3cbe876ab38b7a8c11757d2e90eee2bccbd0)
- formula for eccentricity $e$ in terms of flattening $f$ was corrected (in _ellipsoids.py_):
now it is $e=\sqrt{f*(2-f)}$ (it was $e=\sqrt{f*(1-f)}$)
## Change 2 (088daea24575ceb319c43172737a0c9c87a8b61c)
- avoid unnecessary back-and-forth conversion from DEG to RAD and vice-versa and replace _numpy_ functions with those from _math_ where calculations are performed with scalars (for _wrap_latitude()_ and _wrap_longitude()_ functions in _utils.py_)
- this was done to reduce execution time
```Python
# execution time for current wrap_longitude()
>>> timeit.timeit(stmt="wrap_longitude(pi, radians=True)", setup=setup_wrap_longitude_current, number=1000000)*1000
1109.207500001503

# execution time for new wrap_longitude()
>>> timeit.timeit(stmt="wrap_longitude(pi, radians=True)", setup=setup_wrap_longitude_new, number=1000000)*1000
251.31340000007185
```

```Python
# execution time for current wrap_latitude()
>>> timeit.timeit(stmt="wrap_latitude(pi, radians=True)", setup=setup_wrap_latitude_current, number=1000000)*1000
2117.7677000000585

# execution time for new wrap_latitude()
>>> timeit.timeit(stmt="wrap_latitude(pi, radians=True)", setup=setup_wrap_latitude_new, number=1000000)*1000
480.33010000017384
```

## Change 3 (bacc2481d97cb7f168e9e0ff05b419df870b0d10)
- changes in expressions for converting common latitude to the authalic one and vice versa are based on research by Charles F. F. Karney ([https://doi.org/10.48550/arXiv.2212.05818](https://doi.org/10.48550/arXiv.2212.05818))
- C.F.F. Karney showed that for double-precision floating-point calculations and for ellipsoids with flattening $|f|\le1/150$, when calculating authalic latitude from common latitude, power series expansion with third flattening $n$ as a parameter gives truncation error that is smaller than rounding error (expansion up to $n^6$ is enough for double-precision floating-point calculations)
  - power series approximation gives more accurate results than the direct formula (visible in examples below)
- for the inverse case (authalic to common latitude), the currently used power series with eccentricity squared $e^2$ as a parameter is replaced with the power series that uses third flattening $n$ as a parameter (expansion in $n$ is superior to epansion in $e^2$, as shown by C.F.F. Karney)
- below is the comparison of the obtained results with the current _auth_lat()_ function in the _utils.py_ and a new one
  - calculator on [WolframAlpha.com](https://www.wolframalpha.com/), which performs arbitrary precision calculations, was used to calculate authalic latitude from common using a **direct** formula, and these results were used as a reference for comparisons below

### Comparison of accuracy (for GRS 80 ellipsoid)
#### Radians
|  | $\varphi = 0.1°$ | $\varphi = 60°$ | $\varphi = 89.9°$ |
|--------|--------|--------|--------|
| **$\xi$ (current)** | `0.0017375278003034103` | `1.0452564932058235` | `1.5690431663389075` |
| **$\xi$ (WolframAlpha)** | `0.0017375278003035717413...` | `1.0452564932058239452...` | `1.5690431663389313975...` |
| **$\xi$ (new)** | `0.0017375278003035717` | `1.045256493205824` | `1.5690431663389315` |

$\varphi$ - (common) latitude, 
$\xi$ - authalic latitude

#### Degrees
|  | $\varphi = 0.1°$ | $\varphi = 60°$ | $\varphi = 89.9°$ |
|--------|--------|--------|--------|
| **$\xi$ (current)** | `0.09955300974403514` | `59.888785569338495` | `89.8995513050626` |
| **$\xi$ (WolframAlpha)** | `0.09955300974404437577...` | `59.888785569338517723...` | `89.89955130506396179...` |
| **$\xi$ (new)** | `0.09955300974404438` | `59.888785569338516` | `89.89955130506398` |

$\varphi$ - (common) latitude, 
$\xi$ - authalic latitude

### Comparison of execution times
```Python
# execution time for current auth_lat
>>> timeit.time it(stmt="auth_lat(phi=pi, e=0.08181919104281579, radians=True)", setup=setup_auth_lat_current, number=1000000)*1000
628.6613000002035

# execution time for new auth_lat
>>> timeit.timeit(stmt="auth_lat(phi=pi, e=0.08181919104281579, radians=True)", setup=setup_auth_lat_new, number=1000000)*1000
407.4808000000303

# execution time if in current auth_lat() all numpy functions are replaced with corresponding functions from math
>>> timeit.timeit(stmt="auth_lat(phi=pi, e=0.08181919104281579, radians=True)", setup=setup_auth_lat_current_math, number=1000000)*1000
164.26619999992909
```

ChocopieKewpie commented:
these changes look good- what do you think @araichev @ndemaio ?

ndemaio commented:
Having cloned the PR branch and looked at it at my end I proceeded to run the tests. Result: the unit tests succeed while the doctests fail. @fgilic I'd like a second opinion on this, do they fail for you as well?

Other than that, here's what I think.

Change 1:
Good catch! Thank you for that.

Change 2:
Completely agree with the desire to eliminate the conversion to radians and back to degrees. But they way you've done it introduces more code duplication than I'd like. I'd suggest something like this (in the case of wrap_longitude) to make the functions easier to maintain:

`def wrap_longitude(lam: float, radians: bool = False) -> float:
    '''[...documentation, doctests]...'''

    if radians:
        half_range = pi
    else:
        half_range = 180

    if lam <  -half_range or lam >= half_range:
        result = lam % (2 * half_range)
        if result >= half_range:
            result = result - 2 * half_range
    else:
        result = lam
    return result
`

And then do something similar inside function wrap_latitude as well.

Change 3:
Same comment about code duplication as for Change 2. The expressions for radians and degrees look very similar with the exact same constants. I believe there's potential to make the code more compact by simply defining something like `x = phi` for radians and `x = phi * pi / 180` for degrees (or a more meaningful name!), and then just using x in the expression describing the power series.

Nice exploration of the performance increases for `math`  vs. `numpy` too. That's probably useful info for us as well.

Change 4:
Thank you for improving the test coverage! And thank you for splitting your changes into four tidy commits that are easy to review.

TL;DR: I'd really like to approve this PR. But can't until the question about the failing doctests is resolved. I'm attaching a file with the console output from the failed run.

[doctest_failures.txt](https://github.com/user-attachments/files/16334128/doctest_failures.txt)

fgilic commented:
Thank you very much @ndemaio (and @ChocopieKewpie, @alpha-beta-soup) for taking a look at these changes and for clear and educative feedback. Sorry it took me this long to respond.

- I have incorporated suggestions regarding Changes 2 and 3 (code duplication has been removed in _wrap_longitude()_, _wrap_latitude()_, and _auth_lat()_ functions in _utils.py_), comparison of execution times:

```
Exec. time (wrap_longitude() master branch, degrees) = 1421.292099999846
Exec. time (wrap_longitude() master branch, radians) = 1111.1386000011407
Exec. time (wrap_longitude() PR branch, degrees) = 139.1304000007949
Exec. time (wrap_longitude() PR branch, radians) = 263.25550000001385
Exec. time (wrap_longitude() PR branch - new commit, degrees) = 174.96489999939513
Exec. time (wrap_longitude() PR branch - new commit, radians) = 245.84409999988566

Exec. time (wrap_latitude() master branch, degrees) = 1598.9090000002761
Exec. time (wrap_latitude() master branch, radians) = 1589.2768000012438
Exec. time (wrap_latitude() PR branch, degrees) = 282.1368999993865
Exec. time (wrap_latitude() PR branch, radians) = 461.0367999994196
Exec. time (wrap_latitude() PR branch - new commit, degrees) = 319.0702000010788
Exec. time (wrap_latitude() PR branch - new commit, radians) = 475.67349999917496
```
- code in _auth_lat()_ function in _utils.py_ was further optimized by applying [Horner's method](https://en.wikipedia.org/wiki/Horner%27s_method) for calculating polynomials, comparison of execution times:
```
Exec. time (auth_lat() master, degrees) = 727.3406999993313
Exec. time (auth_lat() master, radians) = 608.9518000007956
Exec. time (auth_lat() PR branch, degrees) = 449.1496000009647
Exec. time (auth_lat() PR branch, radians) = 425.87360000106855
Exec. time (auth_lat() PR branch - new commit, degrees) = 237.020600001415
Exec. time (auth_lat() PR branch - new commit, radians) = 237.39540000133275
```
- all above changes are in a single commit, I should have split them in two...sorry
- regarding the issue with failing doctests, they are failing at my end as well, I forgot to run them 🤭
- doctests that are failing are failing because of the changes in the _auth_lat()_ function in _utils.py_ as shown in more detail below
- after these changes, all doctests are passing

> [!IMPORTANT]
> While I was looking at failing test, I noticed a similar issue as the one in Change 1, this time in line 110 in _ellipsoids.py_ (eccentricity is defined as $e = \sqrt{f*(1-f)}$ and should be $e = \sqrt{f*(2-f)}$). I ignored this for now since, when this is corrected, two unittests and multiple doctests are failing. This is an issue only in cases when Ellipsoid object is initialized without defining parameters `b` and/or `e` (but `a` and `f` only).

<details><summary>Details on making doctests pass</summary>
<p>

To check whether doctest failings are related only to the new auth_lat() function, I used an ellipsoid with the same parameters as the one from the failing doctests (eccentricity e = 0.057806308840112 and authalic radius R_A = 6374581.467096525, which turns out are not WGS 84 parameters because of the issue in line 110 in ellipsoids.py) and followed this procedure:
<ol>
<li>take latitude from the expected results</li>
<li>convert latitude from expected results to authalic latitude by direct formula using an arbitrary precision calculator on <a href="https://www.wolframalpha.com" target="_blank">WolframAlpha.com</a></li>
<li>calculate latitude with the code from the PR branch</li>
<li>convert calculated latitude from the previous step to the authalic latitude by direct formula using an arbitrary precision calculator on <a href="https://www.wolframalpha.com" target="_blank">WolframAlpha.com</a></li>
<li>calculate (authalic) latitude on the authalic sphere (R_A = 6374581.467096525; in this step, body of auth_lat() function isn't executed since e = 0 and thus there are no differences between results obtained from code in master branch and in PR one)</li>
<li>compare results (mainly from 2, 4 , and 5)</li>
</ol>

<h3>Test failed 1 (introduction.rst, changes in lines 261-273):</h3>
<code>File "docs/source/introduction.rst", line 275, in introduction.rst
Failed example:
    for i, row in enumerate(cells):
        for j, cell in enumerate(row):
            print(cell, cell.ellipsoidal_shape(), assert_allclose(cell.nucleus(plane=False), expected_results[i][j], rtol=1e-15, atol=0) == None)
Exception raised:
    Traceback (most recent call last):
      File "C:\Users\fgilic\AppData\Local\miniforge3\envs\dggs\Lib\doctest.py", line 1357, in __run
        exec(compile(example.source, filename, "single",
      File "<doctest introduction.rst[33]>", line 3, in <module>
        print(cell, cell.ellipsoidal_shape(), assert_allclose(cell.nucleus(plane=False), expected_results[i][j], rtol=1e-15, atol=0) == None)
                                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "C:\Users\fgilic\AppData\Local\miniforge3\envs\dggs\Lib\site-packages\numpy\testing\_private\utils.py", line 1504, in assert_allclose
        assert_array_compare(compare, actual, desired, err_msg=str(err_msg),
      File "C:\Users\fgilic\AppData\Local\miniforge3\envs\dggs\Lib\contextlib.py", line 81, in inner
        return func(*args, **kwds)
               ^^^^^^^^^^^^^^^^^^^
      File "C:\Users\fgilic\AppData\Local\miniforge3\envs\dggs\Lib\site-packages\numpy\testing\_private\utils.py", line 797, in assert_array_compare
        raise AssertionError(msg)
    AssertionError:
    Not equal to tolerance rtol=1e-15, atol=0
    Mismatched elements: 1 / 2 (50%)
    Max absolute difference: 1.02019726e-10
    Max relative difference: 1.74480149e-12
     x: array([5.088887e-14, 5.847068e+01])
     y: array([5.088887e-14, 5.847068e+01])</code>
</br></br>
<table class="tg"><thead>
  <tr>
    <th class="tg-fymr">Cell</th>
    <th class="tg-0pky"><span style="font-weight:bold">Phi of cell nuclei</span><br>Expected lat. (from doctest)<br>Expected lat. converted to authalic (WolframAlpha)<br>Calculated lat. from code in the PR branch (DGGS based on ellipsoid)<br>Latitude from the PR branch converted to authalic (WolframAlpha)<br>Calculated latitude on the authalic sphere (DGGS based on authalic sphere; the same from PR and master branch)</th>
  </tr></thead>
<tbody>
  <tr>
    <td class="tg-0pky">N2</td>
    <td class="tg-0pky">58.47067782962734<br>58.413661903369945272749208846254449829744087313905011375408727240...<br>58.47067782972936<br>58.413661903472068295108305786378864562905912970802137715095818895...<br>58.41366190347207</td>
  </tr>
  <tr>
    <td class="tg-0pky">N1</td>
    <td class="tg-0pky">58.47067782962734<br>58.413661903369945272749208846254449829744087313905011375408727240<br>58.47067782972936<br>58.413661903472068295108305786378864562905912970802137715095818895<br>58.41366190347207</td>
  </tr>
  <tr>
    <td class="tg-0pky">N0</td>
    <td class="tg-0pky">58.47067782962736<br>58.413661903369965292945711315687540140475431358492924432029272192...<br>58.470677829729375<br>58.413661903472083310255682638560129904948552178075862224059148909...<br>58.41366190347208</td>
  </tr>
  <tr>
    <td class="tg-0pky">Q0</td>
    <td class="tg-0pky">26.438744923100096<br>26.387799960426015951701843748884845451362960497783040710746342891...<br>26.438744923918186<br>26.387799961243003680954112051853415091623002077502188022455707508...<br>26.387799961243005</td>
  </tr>
  <tr>
    <td class="tg-0pky">Q1</td>
    <td class="tg-0pky">26.438744923100096<br>26.387799960426015951701843748884845451362960497783040710746342891...<br>26.438744923918186<br>26.387799961243003680954112051853415091623002077502188022455707508...<br>26.387799961243005</td>
  </tr>
  <tr>
    <td class="tg-0pky">Q2</td>
    <td class="tg-0pky">26.438744923100096<br>26.387799960426015951701843748884845451362960497783040710746342891...<br>26.438744923918186<br>26.387799961243003680954112051853415091623002077502188022455707508...<br>26.387799961243005</td>
  </tr>
  <tr>
    <td class="tg-0pky">R0</td>
    <td class="tg-0pky">26.438744923100096<br>26.387799960426015951701843748884845451362960497783040710746342891...<br>26.438744923918186<br>26.387799961243003680954112051853415091623002077502188022455707508...<br>26.387799961243005</td>
  </tr>
  <tr>
    <td class="tg-0pky">Q3</td>
    <td class="tg-0pky">3.560649871414923e-15<br>3.5527115721110361735259392825994764326651058981816303105103... × 10^-15<br>3.560649871661821e-15<br>3.5527115723573837261940447233928992362364925363287557694924... × 10^-15<br>3.5527115723573832e-15</td>
  </tr>
  <tr>
    <td class="tg-0pky">Q4</td>
    <td class="tg-0pky">3.560649871414923e-15<br>3.5527115721110361735259392825994764326651058981816303105103... × 10^-15<br>3.560649871661821e-15<br>3.5527115723573837261940447233928992362364925363287557694924... × 10^-15<br>3.5527115723573832e-15</td>
  </tr>
  <tr>
    <td class="tg-0pky">Q5</td>
    <td class="tg-0pky">3.560649871414923e-15<br>3.5527115721110361735259392825994764326651058981816303105103... × 10^-15<br>3.560649871661821e-15<br>3.5527115723573837261940447233928992362364925363287557694924... × 10^-15<br>3.5527115723573832e-15</td>
  </tr>
  <tr>
    <td class="tg-0pky">R3</td>
    <td class="tg-0pky">3.560649871414923e-15<br>3.5527115721110361735259392825994764326651058981816303105103... × 10^-15<br>3.560649871661821e-15<br>3.5527115723573837261940447233928992362364925363287557694924... × 10^-15<br>3.5527115723573832e-15</td>
  </tr>
</tbody></table>

Conclusion: by comparing the second, fourth and the last value in each row, it is visible that latitudes obtained with code from the PR branch are closer to the values obtained with arbitrary precision calculation; therefore, the third value in each row can be used to update this doctest.

<h3>Test failed 2 (dggs.py, changes in lines 95-107):</h3>
<code>File "C:\Users\fgilic\PycharmProjects\rhealpixdggs-py\rhealpixdggs\dggs.py", line 109, in dggs
Failed example:
    for i, row in enumerate(cells):
        for j, cell in enumerate(row):
            print(cell, assert_allclose(cell.nucleus(plane=False), expected_results[i][j], rtol=1e-15, atol=0) == None)
Exception raised:
    Traceback (most recent call last):
      File "C:\Users\fgilic\AppData\Local\miniforge3\envs\dggs\Lib\doctest.py", line 1357, in __run
        exec(compile(example.source, filename, "single",
      File "<doctest dggs[13]>", line 3, in <module>
        print(cell, assert_allclose(cell.nucleus(plane=False), expected_results[i][j], rtol=1e-15, atol=0) == None)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "C:\Users\fgilic\AppData\Local\miniforge3\envs\dggs\Lib\site-packages\numpy\testing\_private\utils.py", line 1504, in assert_allclose
        assert_array_compare(compare, actual, desired, err_msg=str(err_msg),
      File "C:\Users\fgilic\AppData\Local\miniforge3\envs\dggs\Lib\contextlib.py", line 81, in inner
        return func(*args, **kwds)
               ^^^^^^^^^^^^^^^^^^^
      File "C:\Users\fgilic\AppData\Local\miniforge3\envs\dggs\Lib\site-packages\numpy\testing\_private\utils.py", line 797, in assert_array_compare
        raise AssertionError(msg)
    AssertionError:
    Not equal to tolerance rtol=1e-15, atol=0
    Mismatched elements: 1 / 2 (50%)
    Max absolute difference: 1.02012621e-10
    Max relative difference: 1.74467997e-12
     x: array([ 0.      , 58.470678])
     y: array([ 0.      , 58.470678])
</code></br>
Conclusion: although generated planar DGGSs in dggs.py and in introduction.rst (from previous test) are different (in introduction.rst it is (north_square=0, south_square=0, N_side=3), and in dggs.py it is (north_square=1, south_square=2, N_side=3)), their ellipsoidal versions are the same. Therefore, this doctest can be updated in the same way and with the same values as the previous one.


<h3>Test failed 3 (dggs.py, changes in lines 2036-2039):</h3>
<code>File "C:\Users\fgilic\PycharmProjects\rhealpixdggs-py\rhealpixdggs\dggs.py", line 2034, in dggs.Cell.vertices
Failed example:
    for p in c.vertices(plane=False):
        print(my_round(p, 13))
Expected:
    (89.9999999999999, 74.3906909487906)
    (120.0, 41.8738577422094)
    (90.0, 41.8738577422094)
    (60.0, 41.8738577422094)
Got:
    (89.9999999999999, 74.390690948837)
    (120.0, 41.8738577425777)
    (90.0, 41.8738577425777)
    (60.0, 41.8738577425777)</code></br>
</br>
DGGS remains the same as in the first failing test (north_square=0, south_square=0, N_side=3, based on WGS84_ELLIPSOID).</br>
</br>
<table class="tg"><thead>
  <tr>
    <th class="tg-0pky">Vertex</th>
    <th class="tg-0pky"><span style="font-weight:bold">Phi of cell vertices</span><br>Expected lat. (from doctest)<br>Expected lat. converted to authalic (WolframAlpha)<br>Calculated lat. from code in the PR branch (DGGS based on ellipsoid)<br>Latitude from the PR branch converted to authalic (WolframAlpha)<br>Calculated latitude on the authalic sphere (DGGS based on authalic sphere; the same from PR and master branch)</th>
  </tr></thead>
<tbody>
  <tr>
    <td class="tg-fymr">1</td>
    <td class="tg-0pky">74.3906909487906<br>74.357528986954207367045551084636167567093973964574257720013442628...<br>74.390690948837<br>74.357528987000695963482950568571466936829084446502224131894109510...<br>74.35752898700069</td>
  </tr>
  <tr>
    <td class="tg-fymr">2</td>
    <td class="tg-0pky">41.8738577422094<br>41.810314895410404103965979305512986630578193395475021958215653318...<br>41.87385774257768<br>41.810314895778593839370234909877784931348860354405625140647201715...<br>41.8103148957786</td>
  </tr>
  <tr>
    <td class="tg-fymr">3</td>
    <td class="tg-0pky">41.8738577422094<br>41.810314895410404103965979305512986630578193395475021958215653318...<br>41.87385774257768<br>41.810314895778593839370234909877784931348860354405625140647201715...<br>41.8103148957786</td>
  </tr>
  <tr>
    <td class="tg-fymr">4</td>
    <td class="tg-0pky">41.8738577422094<br>41.810314895410404103965979305512986630578193395475021958215653318...<br>41.87385774257768<br>41.810314895778593839370234909877784931348860354405625140647201715...<br>41.8103148957786</td>
  </tr>
</tbody></table>
Conclusion: by comparing the second, fourth and the last value in each row, it is visible that latitudes obtained with code from the PR branch are closer to the values obtained with arbitrary precision calculation; therefore, the third value in each row can be used to update this doctest.

<h3>Test failed 4 (dggs.py, changes in lines 2042-2044):</h3>
<code>File "C:\Users\fgilic\PycharmProjects\rhealpixdggs-py\rhealpixdggs\dggs.py", line 2040, in dggs.Cell.vertices
Failed example:
    for p in c.vertices(plane=False, trim_dart=True):
        print(my_round(p, 13))
Expected:
    (89.9999999999999, 74.3906909487906)
    (120.0, 41.8738577422094)
    (60.0, 41.8738577422094)
Got:
    (89.9999999999999, 74.390690948837)
    (120.0, 41.8738577425777)
    (60.0, 41.8738577425777)</code></br></br>
Conclusion: the only difference between this test and the previous one is that in this one is that non-vertex point for dart cell is removed; therefore this doctest can be updated with the identical latitude values as values from the previous one.

<h3>Test failed 5 (dggs.py, changes in lines 2049-2052):</h3>
<code>File "C:\Users\fgilic\PycharmProjects\rhealpixdggs-py\rhealpixdggs\dggs.py", line 2047, in dggs.Cell.vertices
Failed example:
    for p in c.vertices(plane=False):
        print(my_round(p, 14))
Expected:
    (149.99999999999997, -41.87385774220941)
    (-180.0, -41.87385774220941)
    (-150.0, -41.87385774220941)
    (-180.0, -74.39069094879065)
Got:
    (149.99999999999997, -41.87385774257768)
    (-180.0, -41.87385774257768)
    (-150.0, -41.87385774257768)
    (-180.0, -74.39069094883702)</code></br></br>
<table class="tg"><thead>
  <tr>
    <th class="tg-0pky">Vertex</th>
    <th class="tg-0pky"><span style="font-weight:bold">Phi of cell vertices</span><br>Expected lat. (from doctest)<br>Expected lat. converted to authalic (WolframAlpha)<br>Calculated lat. from code in the PR branch (DGGS based on ellipsoid)<br>Latitude from the PR branch converted to authalic (WolframAlpha)<br>Calculated latitude on the authalic sphere (DGGS based on authalic sphere; the same from PR and master branch)</th>
  </tr></thead>
<tbody>
  <tr>
    <td class="tg-fymr">1</td>
    <td class="tg-0pky">-41.87385774220941<br>-41.81031489541041410151500190381741698491471783446817013605060986...<br>-41.873857742577684<br>-41.81031489577859783838984394931357691618050373853997748633166603...<br>-41.810314895778596</td>
  </tr>
  <tr>
    <td class="tg-fymr">2</td>
    <td class="tg-0pky">-41.87385774220941<br>-41.81031489541041410151500190381741698491471783446817013605060986...<br>-41.873857742577684<br>-41.81031489577859783838984394931357691618050373853997748633166603...<br>-41.810314895778596</td>
  </tr>
  <tr>
    <td class="tg-fymr">3</td>
    <td class="tg-0pky">-41.87385774220941<br>-41.81031489541041410151500190381741698491471783446817013605060986...<br>-41.873857742577684<br>-41.81031489577859783838984394931357691618050373853997748633166603...<br>-41.810314895778596</td>
  </tr>
  <tr>
    <td class="tg-fymr">4</td>
    <td class="tg-0pky">-74.39069094879065<br>-74.35752898695425746251585000644692099840389945694455507280565297...<br>-74.39069094883702<br>-74.35752898700071600167107013733334503813272496681465103576406002...<br>-74.35752898700072</td>
  </tr>
</tbody></table>

Conclusion: by comparing the second, fourth and the last value in each row, it is visible that latitudes obtained with code from the PR branch are closer to the values obtained with arbitrary precision calculation; therefore, the third value in each row can be used to update this doctest.

<h3>Test failed 6 (dggs.py, changes in lines 2055-2057):</h3>
<code>File "C:\Users\fgilic\PycharmProjects\rhealpixdggs-py\rhealpixdggs\dggs.py", line 2053, in dggs.Cell.vertices
Failed example:
    for p in c.vertices(plane=False, trim_dart=True):
        print(my_round(p, 14))
Expected:
    (149.99999999999997, -41.87385774220941)
    (-150.0, -41.87385774220941)
    (-180.0, -74.39069094879065)
Got:
    (149.99999999999997, -41.87385774257768)
    (-150.0, -41.87385774257768)
    (-180.0, -74.39069094883702)</code></br></br>

Conclusion: this doctest can be updated with the identical latitude values as values from the previous one.

<h3>Test failed 7 (dggs.py, changes in lines 960-965):</h3>
<code>File "C:\Users\fgilic\PycharmProjects\rhealpixdggs-py\rhealpixdggs\dggs.py", line 958, in dggs.RHEALPixDGGS.cell_latitudes
Failed example:
    for phi in rdggs.cell_latitudes(1, -pi/2, pi/2, plane=False):
        print(my_round(phi, 14))
Expected:
    -1.02050584399985
    -0.46144314900303
    -0.0
    0.46144314900303
    1.02050584399985
    1.5707963267949
Got:
    -1.02050584400163
    -0.46144314901731
    -0.0
    0.46144314901731
    1.02050584400163
    1.5707963267949</code></br></br>
DGGS is the same as in previous tests, the only difference is that it uses radians instead of degrees as angular unit.</br>
</br>
<table class="tg"><thead>
  <tr>
    <th class="tg-0pky">Nuclei</th>
    <th class="tg-0pky"><span style="font-weight:bold">Phi of cell nuclei</span><br>Expected lat. (from doctest)<br>Expected lat. converted to authalic (WolframAlpha)<br>Calculated lat. from code in the PR branch (DGGS based on ellipsoid)<br>Latitude from the PR branch converted to authalic (WolframAlpha)<br>Calculated latitude on the authalic sphere (DGGS based on authalic sphere; the same from PR and master branch)</th>
  </tr></thead>
<tbody>
  <tr>
    <td class="tg-fymr">1</td>
    <td class="tg-0pky">-1.02050584399985<br>-1.019510728360584174911948861234812728242751401996507403706804440...<br>-1.02050584400163<br>-1.019510728362365972400668647098594000904006857011778621577312378...<br>-1.019510728362366</td>
  </tr>
  <tr>
    <td class="tg-fymr">2</td>
    <td class="tg-0pky">-0.46144314900303<br>-0.460553991667061690569874974890350404266793164349014394719107194...<br>-0.46144314901731004<br>-0.460553991681322490057989438362214893770208024984450792492077633...<br>-0.4605539916813225</td>
  </tr>
  <tr>
    <td class="tg-fymr">3</td>
    <td class="tg-0pky">-0.0<br>0<br>-1.2429012754242461e-16<br>-1.24013028622682197774210622126684516639954524835654800241... × 10^-16<br>-1.240130286226822e-16</td>
  </tr>
  <tr>
    <td class="tg-fymr">4</td>
    <td class="tg-0pky">0.46144314900303<br>0.4605539916670616905698749748903504042667931643490143947191071943...<br>0.4614431490173098<br>0.4605539916813222503813584802165714435272016874998209252723560341...<br>0.4605539916813223</td>
  </tr>
  <tr>
    <td class="tg-0pky">5</td>
    <td class="tg-0pky">1.02050584399985<br>1.0195107283605841749119488612348127282427514019965074037068044407...<br>1.0205058440016297<br>1.0195107283623656720977211100549687163069899012059310596127718932...<br>1.0195107283623657</td>
  </tr>
  <tr>
    <td class="tg-0pky">6</td>
    <td class="tg-0pky">1.5707963267949<br>1.5707963267948932309110705284178269936966586598187817008911183911...<br>1.5707963267948966<br>1.5707963267948965999570432819667607179210729941031869949510934135...<br>1.5707963267948966</td>
  </tr>
</tbody></table>

Conclusion: by comparing the second, fourth and the last value in each row, it is visible that latitudes obtained with code from the PR branch are closer to the values obtained with arbitrary precision calculation; therefore, the third value in each row can be used to update this doctest.


<h3>Test failed 8 (dggs.py, changes in lines 977-982):</h3>
<code>File "C:\Users\fgilic\PycharmProjects\rhealpixdggs-py\rhealpixdggs\dggs.py", line 975, in dggs.RHEALPixDGGS.cell_latitudes
Failed example:
    for phi in rdggs.cell_latitudes(1, -pi/2, pi/2, nucleus=False, plane=False):
        print(my_round(phi, 14))
Expected:
    -1.29836248988994
    -0.73083668811327
    -0.22457715619516
    0.22457715619516
    0.73083668811327
    1.29836248988994
Got:
    -1.29836248989075
    -0.7308366881197
    -0.22457715620807
    0.22457715620807
    0.7308366881197
    1.29836248989075</code></br></br>

<table class="tg"><thead>
  <tr>
    <th class="tg-0pky">Cell boundary</th>
    <th class="tg-0pky"><span style="font-weight:bold">Phi of cell boundary</span><br>Expected lat. (from doctest)<br>Expected lat. converted to authalic (WolframAlpha)<br>Calculated lat. from code in the PR branch (DGGS based on ellipsoid)<br>Latitude from the PR branch converted to authalic (WolframAlpha)<br>Calculated latitude on the authalic sphere (DGGS based on authalic sphere; the same from PR and master branch)</th>
  </tr></thead>
<tbody>
  <tr>
    <td class="tg-fymr">1</td>
    <td class="tg-0pky">-1.29836248988994<br>-1.297783704469474242626376688760080539554575498948980473775333347...<br>-1.2983624898907506<br>-1.297783704470286390390862809916873718350842959959135476197681957...<br>-1.2977837044702865</td>
  </tr>
  <tr>
    <td class="tg-fymr">2</td>
    <td class="tg-0pky">-0.73083668811327<br>-0.729727656220537182863956985475150275159448690417551822010497779...<br>-0.7308366881197008<br>-0.729727656226966406689409594700444078274169245678919295321140634...<br>-0.7297276562269664</td>
  </tr>
  <tr>
    <td class="tg-fymr">3</td>
    <td class="tg-0pky">-0.22457715619516<br>-0.224093092288485871946436980865764045300705348192057463235872215...<br>-0.2245771562080711<br>-0.224093092301371034706982623115426149855266748517933666220181264...<br>-0.22409309230137103</td>
  </tr>
  <tr>
    <td class="tg-fymr">4</td>
    <td class="tg-0pky">0.22457715619516<br>0.2240930922884858719464369808657640453007053481920574632358722150...<br>0.2245771562080707<br>0.2240930923013706355105466712351614150438576682375751445533462475...<br>0.22409309230137064</td>
  </tr>
  <tr>
    <td class="tg-0pky">5</td>
    <td class="tg-0pky">0.73083668811327<br>0.7297276562205371828639569854751502751594486904175518220104977791...<br>0.7308366881197005<br>0.7297276562269661067629389167427597697171330372508046144495482635...<br>0.7297276562269661</td>
  </tr>
  <tr>
    <td class="tg-0pky">6</td>
    <td class="tg-0pky">1.29836248988994<br>1.2977837044694742426263766887600805395545754989489804737753333475...<br>1.2983624898907502<br>1.2977837044702859896271004185416361771430767027144229168154458733...<br>1.297783704470286</td>
  </tr>
</tbody></table>


Conclusion: by comparing second, fourth and the last value in each row, it is visible that latitudes obtained with code from the PR branch are closer to the values obtained with arbitrary precision calculation; therefore, the third value in each row can be used to update this doctest.

<h3>Test failed 9 (projection_wrapper.py, change in line 55):</h3>
<code>File "C:\Users\fgilic\PycharmProjects\rhealpixdggs-py\rhealpixdggs\projection_wrapper.py", line 54, in projection_wrapper.Projection
Failed example:
    print(my_round(f(0, 30), 15))
Expected:
    (0.0, 3748655.1150495014)
Got:
    (0.0, 3748655.115049504)</code></br></br>

<ul>
<li>expected projection of point (0,30) on WGS84_ELLIPSOID ellipsoid using rhealpix projection is (0.0, 3748655.1150495014)</li>
<li>projection of point (0,30) on WGS84_ELLIPSOID ellipsoid using rhealpix projection and code from the PR branch is (0.0, 3748655.115049504)</li>
<li>when common latitude of 30° on WGS84_ELLIPSOID (e = 0.057806308840112; R_A = 6374581.467096525) is converted to authalic latitude using direct formula and arbitrary precision calculation (on WolframAlpha.com), result is: 29.944661525272687580648918634180720123042154571015225487721365461...°</li>
<li>when point (0, 29.9446615252726875806489186...) on the authalic sphere (R = 6374581.467096525) is projected from the authalic sphere to the plane using rhealpix projection (code from PR branch, but in this case body of auth_lat() function is not executed since e = 0), result is (0.0, 3748655.115049504), which is the same as the result that is obtained with code from the PR branch, confirming it is correct</li>
</ul>

Conclusion: this doctest can be updated with the y coordinate 3748655.115049504.

</p>
</details> 

Regarding the issue with defining eccentricity _e_ in line 110 in _ellipsoid.py_, do you suggest this should also be addressed in this PR?
Thanks!

araichev commented:
This PR looks like a useful set of changes for improved accuracy and computational speed.
I haven't dived deep enough to comment on code quality and testing, so take my comment as cheering from the sidelines :)

rggibb commented:
Similar to @araichev take this comment as cheering from the sidelines ;) however I do have a couple of ancilliary comments/questions which may be easily answered or may require new issues to be raised. The first comment is a consequence of this code base preceeding the writting of OGC AS Topic 21, the second comment arises from the work of John Herring on efficient computation of area and length on ellispoidal surfaces.

1. One thing I have been aware of for a long time that we havent implemented is that OGC AS Topic 21 v2/ISO 19170.1 says that every DGGS shall have an error budget as described in [Section 9.1.5.3.  Cells of equal area](https://docs.ogc.org/as/20-040r3/20-040r3.html#req-ea-ers-cell-equal_area). The error budget controls both the precision that the equal area test is computed to and also the maximum practical cell depth that is supported by the DGGS. The error budget is a mixture of both the precision to which the ellipsoid is defined and the limits of precision in the implementation code. Without having looked at it it in great detail, it would seem to me that this work highlights the precision of the implementation and is an opportunity for us to be explicit about the error budget we are using and as a consequence the practical depth ie number of cell tesselation iterations, or smallest area cell that the code and therefore the DGGS supports.
2. John's work was published as [Geographic information — Features and Geometry — Part 2: Measure](https://portal.ogc.org/files/?artifact_id=92693), which isn't a standard but rather an implementation reference for efficient ellipsoidal maths. Knowing John I suspect that he would have only bothered to write this document if it both described an approach that was fundamentally different from the way proj approaches ellipsoidal maths, and was also significantly more efficient, I know John sat down with [Matt Purss](https://github.com/geofizzydrink) to explain it, so we could approach Matt for practical comment on the pros and cons .. but basically I'm just asking if others are aware of John's work?

ndemaio commented:
Thank you for the updates to the PR @fgilic ! Every test passes at my end as well now.

About the new issue in ellipsoids.py you discovered - another good catch. I'd normally say that the scope of a PR should creep as little as possible, which would mean that any newly discovered issues should be handled separately.

But.

We now have two different calculations of the same thing in the code, with one being correct and one being wrong. This could introduce unexpected errors and behaviour until the discrepancy is resolved.

I think it comes down to whether we prioritise a finished partial improvement over an unfinished complete one. My preference would be for the partial improvement, while logging the newly discovered issue as a (high-priority) bug in the repo. @alpha-beta-soup and @ChocopieKewpie what do you think? (I'll merge the PR if you're okay with this approach.)

alpha-beta-soup commented:
Yes, I vote for:
- Merge this PR.
- Make an issue for the eccentricity in `ellipsoids.py`, as its changes can stand alone.
- Make an issue stemming form @rggibb's comment 1 regarding the error budget (this will also help us with the precision we should use for testing, as well as being important in its own right).
